### PR TITLE
Refractor `compute`

### LINKED ISSUE #90693: `compute`'s use of SendPorts is not safe
See https://github.com/dart-lang/sdk/issues/47164#issuecomment-926862699 for context.

/cc @lrhn

### PR DESCRIPTION
**Pull request by jellynoone**
This PR fixes #90693 and makes the `compute` method more efficient by using a single `ReceivePort` instead of 3.

The code changes were inspired by @Irhn 's [comment in a Dart SDK issue](https://github.com/dart-lang/sdk/issues/47164#issuecomment-926862699).

Fixes https://github.com/flutter/flutter/issues/90693

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/foundation/_isolates_io.dart
Comment (jellynoone) [inline packages/flutter/lib/src/foundation/_isolates_io.dart:30]: I would have done `port.first` instead, but the previous iteration of the code was timing the time it takes to close the ports and `port.first` does that implicitly. Thus, the use of `Completer` to maintain the original logic.

**Comment by jellynoone** [comment]
There is also an edge case here that also wasn't handled in the previous iteration of the code, which is that no error will be reported if the entry point fails after the sending the result back to the main `Isolate`.

This is because after receiving the data we close all the ports as such no additional handlers are called.

However, this is only an issue for the development team not the clients / consumers.

I've been thinking a great deal about this and there is no simple or secure way of doing this that would be 100% future-proof. Also it's also impossible to test this because it would not something the caller is responsible for but the developer.

To illustrate when this issue would occur:
```dart
// if this were to be changed
Isolate.exit(configuration.resultPort, <R>[result]);
//  to something like this
configuration.resultPort.send(<R>[result]);
configuration.resultPort.send(<R>[result]); // This send event would not be caught
// or 
configuration.resultPort.send(<R>[result]);
throw Error(); // This error would not be reported
```
The previous implementation also ignored additional success / result events, while reported an additional error. But because `close()` was called on the `ReceivePort` it actually wasn't handled. (I tested this).

To actually force the handlers to be called one would have to delay returning and do something like `await Future.delay(Duration.zero);` before closing the ports. And this STILL isn't guaranteed because the failing code could be some newly added listener or delayed response. I. e.

```dart
configuration.resultPort.send(<R>[result]);
await Future.delay(Duration(seconds: 1), () => throw Error());
```

However, as long as we call `Isolate.exit` and only send one response through the `SendPort` we should be fine. Even if some erroneous listener was added in the user provided callback, since the entire `Isolate` is killed.

Am I overthinking this? For now I'll just add a comment and if anyone has a better solution in mind, please let me know.

**Comment by jellynoone** [comment]
@dnfield Would you mind reviewing this?

File: packages/flutter/lib/src/foundation/_isolates_io.dart
Comment (jellynoone) [inline packages/flutter/lib/src/foundation/_isolates_io.dart:47]: ~~Should this instead be an assertion? We are in control of the Isolate so someone else can't kill it under underneath us. And if the main Isolate is killed so is ours.~~

Never mind, forgot the user can call `Isolate.current.kill` from the provided callback.

File: packages/flutter/lib/src/foundation/_isolates_io.dart
Comment (jellynoone) [reply packages/flutter/lib/src/foundation/_isolates_io.dart]: Changed `ReceivePort` to `RawReceivePort` because of the limited API we are using

**Comment by jellynoone** [comment]
I also added additional tests:
- calling compute from compute callback,
- testing executing `Isolate.exit` from the compute callback, _(Tests the exit branch (which wasn't tested before))_
- and all of the different types of `Isolate.current.kill` i.e. immediate / beforeNextEvent

**Review by dnfield — APPROVED** [review]
This LGTM. Would appreciate a review from @aam or @lrhn as well.

File: packages/flutter/lib/src/foundation/_isolates_io.dart
Comment (dnfield) [inline packages/flutter/lib/src/foundation/_isolates_io.dart:87]:
```suggestion
/// [SendPort.send] be it directly or indirectly i.e. [Isolate.exit].
```

Comment (jellynoone) [reply packages/flutter/lib/src/foundation/_isolates_io.dart]: Fixed and thank you for reviewing this PR.

File: packages/flutter/lib/src/foundation/_isolates_io.dart
Comment (jellynoone) [inline packages/flutter/lib/src/foundation/_isolates_io.dart:98]: I also saw timing in _spawn can be simplified, and applied the fix as well.

Both solutions side by side in a contrived example:
```dart
import 'dart:async';
import 'dart:developer';

void main() async {
  Timeline.timeSync('test', compute);
  print('finished');

  unawaited(Timeline.timeSync<Future<int>>('test', () async {
    final FutureOr<int> applicationResult = await compute();
    return await applicationResult;
  }));
  print('finished');
}

int _count = 0;
FutureOr<int> compute() async {
  final i = fib(40);

  print('DID ${_count++}');

  return i;
}

int fib(int i) {
  if (i < 2) return i;

  return fib(i - 2) + fib(i - 1);
}
```

**Comment by jellynoone** [comment]
@dnfield from the looks of it in [Dart 2.17 a new method `Isolate.run` is added which is essentially the same as `compute`](https://github.com/dart-lang/sdk/commit/05322f23dd600f170bf86278e61071dceac196ca) (in terms of functionality).

The differences I see in the implementation is:
1. we time the functions i.e. `Timeline`
2. we use `Exception` instead of `RemoteError` or the error that was actually thrown by the user provided function
3. I see there is some additional safety around closing the receive port
4. and they use non-growable lists

For now, I will address 3. and 4.
Because 1. and 2. would be a breaking change in `compute`, I will leave those as is.

How should we proceed?
- deprecate compute and lose the `Timeline`
- keep our own version

**Comment by jellynoone** [comment]
After taking a second look at the new `Isolate.run` implementation it does not have exactly the same functionality. 

The `run` method accepts a no arguments function. 
While we require a function with an argument and the argument to be passed to it. 

Which means the transition for end users wouldn't be *just* replacing the call but still relatively trivial if we chose to deprecate `compute`.

Anyways, awaiting additional feedback.

**Comment by dnfield** [comment]
I think it would make sense to do the following:

- Deprecate compute, and point people instead to `Isolate.run`.
- Make its functionality forward to `Isolate.run`. Keep the timeline functionality in `compute`.

@goderbauer @Piinks wdyt?

**Comment by jellynoone** [comment]
If we want to keep timing exactly intact, we need to actually implement the function ourselves. _Since the implementation timed the `ReceivePort` constructing and closing._ 

If we're OK with simply timing the user callback then the new implementation is essentially:

```dart
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:async';
import 'dart:developer';
import 'dart:isolate';

import 'constants.dart';
import 'isolates.dart' as isolates;

/// The dart:io implementation of [isolate.compute].
Future<R> compute<Q, R>(isolates.ComputeCallback<Q, R> callback, Q message, { String? debugLabel }) {
  debugLabel ??= kReleaseMode ? 'compute' : callback.toString();

  return Isolate.run(() => Timeline.timeSync(
    debugLabel!,
    () => callback(message),
  ));
}
```

Of course test expectations would need to be changed + the raised issue in the Dart SDK resolved.

**Comment by goderbauer** [comment]
My preferred way would be to keep the compute functionality, but implement it in terms of the new Dart API. The tracing that compute adds over the raw Dart API has been useful. The important bits to keep for tracing here are the flow events (going from the point the computation was kicked of to the computation to the point where the result of the computation is received) and the duration of the computation. Measuring how long it takes to instantiate and close the ports is not important, IMO. I believe that was just done in the current implementation to get the flow events inserted properly.

I don't think we should deprecate `compute` altogether since it does offer the timeline tracing as an additional value. We should adjust the documentation, though, to document this difference over the other API.

Is the switch from Exception to RemoteError actually breaking anything?

**Comment by jellynoone** [comment]
> Is the switch from Exception to RemoteError actually breaking anything?

@goderbauer It breaks tests and if this is fine, I have no problem changing the tests.

Personally, I do prefer the approach the `Isolate.run` is taking which is actually forwarding the errors and exceptions from the `Isolate.run` function back to the main isolate.

So, for instance, I can do:

```dart
try {
  Isolate.run(() => throw MyCustomException);
} on MyCustomException {
  print("caught");
}

// Prints "caught"
```

With our implementation of `compute` you could only catch an `Exception`. Which isn't that helpful and probably counter-intuitive.

```dart
try {
  compute((_) => throw MyCustomException, null);
} on MyCustomException {
  print("caught");
}

// Throws Uncaught Exception
```

I also think, we should keep all the tests we have, even the new ones added, since they also helped me catch a bug in the implementation of `Isolate.run`.

I was also wondering are we constrained by a specific Dart SDK version i.e. is it fine to already use this functionality or would we release first this version of `compute` and later when the minimum Dart SDK version increases the new one?
_I'm a new contributor so I don't know how this is done._

I already have a running and passing local changes for a patch that uses `Isolate.run`.  Should I open a new PR or just change the title here?

**Comment by goderbauer** [comment]
> I was also wondering are we constrained by a specific Dart SDK version i.e. is it fine to already use this functionality or would we release first this version of compute and later when the minimum Dart SDK version increases the new one?

Has the version of the Dart SDK that provides this functionality already rolled into Flutter? If so, it's ok to use it.

The approach to errors that Isolate.run is taking does sound a little nicer. If it is just breaking our own tests, I think it would be fine to just switch to that.

**Comment by jellynoone** [comment]
`Isolate.run` comes with Dart 2.17 which hasn't even released yet. And if I understand:

> Has the version of the Dart SDK that provides this functionality already rolled into Flutter? 

correctly.

Currently, only features of Dart 2.12 can be used?

**Comment by goderbauer** [comment]
The dart version that's currently bundled with Flutter on `master` is `2.17.0-182.0.dev`. So everything included in that version can be used in the framework.

**Comment by jellynoone** [comment]
Oh!!! Wonderful!

I was able to make the required changes. 

Two notes:
1. I improved some comments. With Dart 2.15 it became possible to send closures directly across isolates as well as instance methods. I also added tests to ensure these assumptions stay true in the future.
2. Until the [`Isolate.run` bug](https://github.com/dart-lang/sdk/issues/48516) is fixed one test is marked for skipping.

File: packages/flutter/lib/src/foundation/isolates.dart
Comment (dnfield) [inline packages/flutter/lib/src/foundation/isolates.dart:61]: We should just have a line under See also about `Isolate.run`. We should not document implementation details here. If anything, we could say that this method works similarly to `Isolate.run` with additional tracing.

**Review by dnfield — COMMENTED** [review]
This is great. One more comment about documentation.

My only other question is whether we should wait to land this until the Dart VM fix lands and rolls into the framework, so we don't have to temporarily disable the test and remember to enable it.

File: packages/flutter/lib/src/foundation/isolates.dart
Comment (jellynoone) [reply packages/flutter/lib/src/foundation/isolates.dart:61]: Would just adding:

```dart
/// See also:
/// ...
/// * [Isolate.run], for an alternative version of this function without tracing information
```

be fine?

File: packages/flutter/lib/src/foundation/isolates.dart
Comment (goderbauer) [inline packages/flutter/lib/src/foundation/isolates.dart:58]: nit: sentences in docs should start with a capital letter. You may have to restructure this sentence a little bit so it doesn't start with lower-case `compute`.

Comment (jellynoone) [reply packages/flutter/lib/src/foundation/isolates.dart:58]: @dnfield recommended I just add a section to `See also` and skip documenting the internals all together

Comment (goderbauer) [reply packages/flutter/lib/src/foundation/isolates.dart:58]: That sounds good, too. Sorry I didn't see that comment.

**Comment by jellynoone** [comment]
I will land the patch addressing the reviews, tomorrow. _It's almost midnight here._

I also want to wait for all the tests to pass and see if the difference in error handling broke anything.

I too suspect that the fix for bug with `Isolate.run` will land soon considering [the patch fixing it is already written](https://dart-review.googlesource.com/c/sdk/+/235603). I think we should wait so I also push a commit not skipping the failing test.

**Comment by jellynoone** [comment]
@goderbauer can you check if the failure is related to this patch?

**Comment by jellynoone** [comment]
OK, thank you for letting me know.

I think we should also then wait for the fix for `Isolate.run` bug to roll into the framework, which I think will happen by the end of the week. Then, I'll also unskip that buggy test.

Unless we are nearing a deadline in which case I case and can push the commit with addressed changes. (The reason I haven't is because I don't want to needlessly waste resources. (For a comment fix of all things))

**Comment by jellynoone** [comment]
The [fix for bug in `Isolate.run`](https://github.com/dart-lang/sdk/commit/33c38e914bff88bf39d488ee9c1a3c0613a3a30c) landed in [2.17.0-193.0.dev](https://github.com/dart-lang/sdk/releases/tag/2.17.0-193.0.dev).

@goderbauer Any idea when that version will be rolled into Flutter?

**Comment by dnfield** [comment]
@jellynoone the Dart roller is currently paused for branch alignment - it should be turned back on again before long (i.e. latest should be sometime early next week). Once it's back on it will roll Dart into flutter/engine, and then that needs to be rolled into framework.

I would expect that we should be able to land this sometime in the next week or two.

**Comment by jellynoone** [comment]
Great, thank you for informing me.

I'll keep checking and push when the fix rolls into Dart.

**Comment by goderbauer** [comment]
> Could we do the same in compute?

This would be a pretty big breaking change - and I don't see any easy non-breaking migration path. If we wanted to do that, we'd have to deprecate the current `compute` and implement a new one (finding a proper name for that one will be hard).

I think, if somebody wants the simplified semantics, they can just use `Isolate.run` directly.

**Comment by goderbauer** [comment]
FYI: I believe `Isolate.run` has some usability issues because the closure provided lents itself to over-capturing stuff that cannot be send to an isolate. See https://github.com/dart-lang/sdk/issues/48566.

**Comment by jellynoone** [comment]
But isn't this also an issue with `compute` now, since arbitrary closures can be passed to the function and to the best of my knowledge we really have no way of knowing if the function is static / top-level.

Except, if we do a very implementation depended [`assert(callback.toString().contains(': static'))`](https://github.com/dart-lang/sdk/blob/8c484d12fe8973092f84a5ae1b4fc2273774de68/runtime/vm/object.cc#L10148)

**Comment by goderbauer** [comment]
Yes, you can also run into this problem with `compute`, but the design of `Isolate.run` makes it more likely that you trigger this case, I think, because the only way to pass in data is to closure it in.

**Comment by jellynoone** [comment]
I explicitly changed comments in this PR to be more inline with the current dart version features. i.e. made it explicit that it is possible to send closure as apposed to just top-level and static functions.

Should I add another comment recommending sticking to just top-level and static functions? Citing this problem or simply adding a case for easier testing?

**Comment by goderbauer** [comment]
It may be useful to include a link to the underlying issue https://github.com/dart-lang/sdk/issues/36983 and inform people that closures in dart may capture more than meets the eye.

File: packages/flutter/lib/src/foundation/_isolates_io.dart
Comment (jellynoone) [inline packages/flutter/lib/src/foundation/_isolates_io.dart:31]: Would it make sense to have different ends label for a failed computation and a successful one?

**Comment by jellynoone** [comment]
Addressed previous comments, added note about closures and made sure the `Flow.end` event is called even when the computation fails. (Missed this before).

**Comment by jellynoone** [comment]
As stated in [dart-lang/sdk#48566](https://github.com/dart-lang/sdk/issues/48566#issuecomment-1070619047), a decision was made to *NOT* ship `Isolate.run` in the next stable release.

I re-implemented the improved `compute` we had before with a few notable changes to bring it more inline with how `Isolate.run` did things:
- errors are now caught and sent back to the main Isolate and throw,
- if the isolate exited prematurely a `RemoteError` is thrown instead of an `Exception`,
- it now sets the `Isolate.debugName`

additional changes:
- I took the liberty to remove one of the previous timing functions which really did nothing and was already removed in the version which used `Isolate.run`,
- `_IsolateConfiguration` now also times the `callback`,

**Review by dnfield — APPROVED** [review]
Still LGTM.

Thanks for sticking with this. It may not be apparent but your work here has helped improve things around the `Isolate.run` API and identify some key areas needed to improve that.

**Comment by dnfield** [comment]
This still needs a second LGTM - @goderbauer?

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 24.
View them at https://flutter-gold.skia.org/cl/github/99527

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/foundation/_isolates_io.dart
Comment (aam) [inline packages/flutter/lib/src/foundation/_isolates_io.dart:69]: is this redundant considering we do explicit `as` on the next line?

(same question to three asserts below)

**Review by aam — APPROVED** [review]

File: packages/flutter/lib/src/foundation/_isolates_io.dart
Comment (jellynoone) [reply packages/flutter/lib/src/foundation/_isolates_io.dart:69]: I added them because I wanted it to be clear it is our job to guarantee the type is `R`.

While I agree that they are redundant, I would assume if someone else read the code they wouldn't have to wonder and look at the surrounding code to make sure the `... as Type` casts are safe.

I also took for inspiration [the initial comment by @irhn in a dart sdk issue](https://github.com/dart-lang/sdk/issues/47164#issuecomment-926862699) where the assert was used to assert the response was a `List<dynamic>`. Which could also be argued in this pull request as redundant since an `as` cast follows it immediately.

I'm not personally attached and can remove these `assert`s but if possible I would prefer to keep them since *I think* they make the code clearer for a new reader (at least the intent).

**Comment by jellynoone** [comment]
> Still LGTM.
> 
> Thanks for sticking with this. It may not be apparent but your work here has helped improve things around the `Isolate.run` API and identify some key areas needed to improve that.

@dnfield  Thank you very much. It was a fun exercise.

File: packages/flutter/lib/src/foundation/_isolates_io.dart
Comment (aam) [reply packages/flutter/lib/src/foundation/_isolates_io.dart:69]: > FWIW, Flutter style encourages liberal use of asserts if they can be at all helpful for debugging:

What is mentioned in the doc makes sense, but `assert(v is T)` is built into `v as T`, produces basically exact same exception being thrown.

Not a big deal for sure though.

Comment (jellynoone) [reply packages/flutter/lib/src/foundation/_isolates_io.dart:69]: Hypothetically:

```dart
void main(List<String> args) => args.isEmpty ? fn1(1) : fn2(1);

void fn1(dynamic args) {
  assert(args is List);
}

void fn2(dynamic args) => args as List;
```

running: `dart run --enable-asserts main.dart`  
produces: `... Failed assertion: line 4 pos 10: 'args is List': is not true. ...`

while `dart run --enable-asserts main.dart 1` 
produces: `... type 'int' is not a subtype of type 'List<dynamic>' in type cast ...`

Again, personally I prefer the assertion during development.
Because a type cast signals to me that I might not understand the program flow correctly or that I made the wrong assumption.
While an assertion to me means this was intended and that something somewhere changed which made this intended assumption incorrect.

*Also I always feel a bit anxious if I don't use an assertion before a type cast.* Even in cases like this where the code guaranteeing the assumption is close. But that might easily change in a future change.

An additional reason why I added the assertions is because I feel the most correct way to write that would have been:
```dart
final dynamic result = response[0];
if(result is R) {
  return result;
} else {
   throw Exception('Should never happen!!!');
}
```
But that code is an assertion! 
Which could be reduced to:
```dart
assert(response[0] is R);
return response[0];
```

But because the type of `response[0]` is inferred as `dynamic`. We need to add the explicit cast, and arrive at the code in the PR.

I feel, I'm making a big deal out of this rival detail. But just wanted to explain how I reasoned.

**Comment by jellynoone** [comment]
All tests are ~~green~~ some broken again, and I just wanted to add one last note for documentation purposes of the changes.

User facing changes (tiny):
- one (redundant) timing event was removed;
- the spawned Isolate now named;

User facing changes (minor):
- Exceptions and errors thrown by the computation are re-thrown in the main `Isolate`;
- Errors thrown in the `Isolate` must be sendable between `Isolate`s
- `RemoteError` is thrown instead of an `Exception` when: 
1. a user does something freaky like immediately kill or exit the current `Isolate` from the computation;
2. a user sends back a response which cannot be sent through the `SendPort.send`;
3. a user throws an error from the computation which cannot be sent through the `SendPort.send`; or
4. we made a mistake somewhere (should only happen when/if new code is add.

Internal:
- use 1 `RawReceivePort` instead of 3 `ReceivePort`s,
- simplified specified types,
- wrapping spawned `Isolate` result in `List` to easily differentiate between different event types

**Comment by jellynoone** [comment]
@mit-mit I don't know if you're the right person to tag.

I just want to make sure users aren't forgotten to be informed about changes in my previous post (mostly exception handling).

**Comment by lrhn** [comment]
(About the assert: Until all programs are soundly null-safe, `as Foo` is potentially effectively doing `is Foo?`.)

**Comment by jellynoone** [comment]
Aha!!! Thank you for the explanation.

**Pull request merged**
